### PR TITLE
ASoC: topology: disable size checks for bytes_ext controls if needed

### PR DESCRIPTION
**Pull request by plbossart**
When CONFIG_SND_CTL_VALIDATION is set, accesses to extended bytes
control generate spurious error messages when the size exceeds 512
bytes, such as

[ 11.224223] sof_sdw sof_sdw: control 2:0:0:EQIIR5.0 eqiir_coef_5:0:
invalid count 1024

In addition the error check returns -EINVAL which has the nasty side
effect of preventing applications accessing controls from working,
e.g.

root@plb:~# alsamixer
cannot load mixer controls: Invalid argument

It's agreed that the control interface has been abused since 2014, but
forcing a check should not prevent existing solutions from working.

This patch skips the checks conditionally if CONFIG_SND_CTL_VALIDATION
is set and the byte array provided by topology is > 512. This
preserves the checks for all other cases.

Fixes: 1a3232d2f61d2 ('ASoC: topology: Add support for TLV bytes controls')
BugLink: https://github.com/thesofproject/linux/issues/2430
Reported-by: Takashi Iwai <tiwai@suse.de>
Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

### CODE REVIEW

**Comment by bardliao** [comment]
IMHO, this looks a little bit hacky. Can we move the fix to sound/soc/sof/topology.c?
Like
```patch
diff --git a/sound/soc/sof/topology.c b/sound/soc/sof/topology.c
index eaa1122d5a68..1b149b4107eb 100644
--- a/sound/soc/sof/topology.c
+++ b/sound/soc/sof/topology.c
@@ -1167,6 +1167,7 @@ static int sof_control_load_bytes(struct snd_soc_component *scomp,
                ret = -EINVAL;
                goto out;
        }
+       kc->access |= SNDRV_CTL_ELEM_ACCESS_SKIP_CHECK;
 
        scontrol->size = sizeof(struct sof_ipc_ctrl_data) + priv_size;
```
I tested above patch and it works on my side.

**Comment by ranj063** [comment]
> IMHO, this looks a little bit hacky. Can we move the fix to sound/soc/sof/topology.c?
> Like
> 
> ```diff
> diff --git a/sound/soc/sof/topology.c b/sound/soc/sof/topology.c
> index eaa1122d5a68..1b149b4107eb 100644
> --- a/sound/soc/sof/topology.c
> +++ b/sound/soc/sof/topology.c
> @@ -1167,6 +1167,7 @@ static int sof_control_load_bytes(struct snd_soc_component *scomp,
>                 ret = -EINVAL;
>                 goto out;
>         }
> +       kc->access |= SNDRV_CTL_ELEM_ACCESS_SKIP_CHECK;
>  
>         scontrol->size = sizeof(struct sof_ipc_ctrl_data) + priv_size;
> ```
> 
> I tested above patch and it works on my side.

I actually like @plbossart 's implementation. It would apply to all drivers.

**Review by ranj063 — APPROVED** [review]

**Comment by juimonen** [comment]
@plbossart where is the 512 magic value coming from? also is this bytes "ext" somehow different than only "bytes"? Also is this related to this other access control patch just merged in fw 333036f097cad04d8c4514eb736d95fc81bf27b0?

So what I basically am after here is that we have some max control size coming from topology, so are we just not having enough big value defined on some topology file?

**Comment by bardliao** [comment]
> I actually like @plbossart 's implementation. It would apply to all drivers.

Yes, but the question is that do we want it to be applied to all drivers?

**Comment by plbossart** [comment]
> > I actually like @plbossart 's implementation. It would apply to all drivers.
> 
> Yes, but the question is that do we want it to be applied to all drivers?

Answering to both @juimonen and @bardliao 

Historically the bytes control are limited to 512. Back in 2014, the Intel folks suggested an extension with bytes_ext and TLV, where the actual data is not limited and the size of the buffer is provided in the L of the TLV header. 

However it's been a horrible hack. More specifically the part that's questionable is that the TLV can be used with the regular bytes, but we actually always assume in the topology code that the presence of the TLV means an extended access.

Bottom line, I would like to *only* modify the code which is used by the topology. I am pretty sure this issue touches the Skylake driver. The point is really that there is no way to detect the abuse, and whether a regular bytes control is used, or if it's been abused and the length is not limited to 512.

Long term, the direction is to stop using such controls and use a hw-dep interface. I did some work on this maybe 2 years ago, it may be time to revive it and move away from bytes controls.

**Comment by bardliao** [comment]
Thanks @plbossart for the explanation and the change makes sense to me.

**Review by bardliao — APPROVED** [review]

**Comment by juimonen** [comment]
@plbossart ok. Can you still clarify me why this thing is hit now here? I mean we've been sending "big" bytes before also without issues... so some difference in topology etc? Otherwise I'm fine with the change, so just go a head merge.

**Comment by plbossart** [comment]
> @plbossart ok. Can you still clarify me why this thing is hit now here? I mean we've been sending "big" bytes before also without issues... so some difference in topology etc? Otherwise I'm fine with the change, so just go a head merge.

Add CONFIG_SND_CTL_VALIDATION=y to your .config and you'll get the errors.

**Review by juimonen — APPROVED** [review]
@plbossart ok, got it.
